azurerm/internal/services/compute/virtual_machine_scale_set.go

```diff
@@ -1557,9 +1558,44 @@ func VirtualMachineScaleSetExtensionsSchema() *schema.Schema {
 				},
 			},
 		},
+		Set: virtualMachineScaleSetExtensionHash,
 	}
 }
 
+func virtualMachineScaleSetExtensionHash(v interface{}) int {
+	var buf bytes.Buffer
+
+	if m, ok := v.(map[string]interface{}); ok {
+		buf.WriteString(fmt.Sprintf("%s-", m["name"].(string)))
+		buf.WriteString(fmt.Sprintf("%s-", m["publisher"].(string)))
+		buf.WriteString(fmt.Sprintf("%s-", m["type"].(string)))
+		buf.WriteString(fmt.Sprintf("%s-", m["type_handler_version"].(string)))
+		buf.WriteString(fmt.Sprintf("%t-", m["auto_upgrade_minor_version"].(bool)))
+
+		if v, ok = m["force_update_tag"]; ok {
+			buf.WriteString(fmt.Sprintf("%s-", v))
+		}
+
+		if v, ok := m["provision_after_extensions"]; ok {
+			buf.WriteString(fmt.Sprintf("%s-", v))
+		}
+
+		// we need to ensure the whitespace is consistent
+		settings := m["settings"].(string)
+		if settings != "" {
+			expandedSettings, err := structure.ExpandJsonFromString(settings)
+			if err == nil {
+				serializedSettings, err := structure.FlattenJsonToString(expandedSettings)
+				if err == nil {
+					buf.WriteString(fmt.Sprintf("%s-", serializedSettings))
+				}
+			}
+		}
+	}
```

Comment:
Since protected settings isn't returned this means we won't track any changes to that field, which is something we want to avoid - instead we'd need to conditionally add that to the hash here

Reply:
@tombuildsstuff

I've updated the code to cache the existing `extension` settings in state, keyed by the `extension` name. The `extension` name is guaranteed to be unique in Azure today, otherwise, you'll get:

```
Code="BadRequest" Message="On resource 'foobar', extension name 'CustomScript' cannot be used for more than one extension."
```

During the flattening of the `protected_setting`, it will try to set it as long as there is something being set in the state. In v2.60.0, after changing the `extension` from List to Set, the `protected_settings` are not set back to the `ResourceData` during a Read/Import, while it **is** set into the state during the create/update (persumably because the `Read` got invoked during these operations will will read from source of **config**, which ends up being persisted into the state). Therefore, this fix should the state resulted into running the provisioning using v2.60.0 (verified locally).